libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -229,32 +229,53 @@ bool IRGeneratorForStatements::visit(Conditional const& _conditional)
 bool IRGeneratorForStatements::visit(Assignment const& _assignment)
 {
 	_assignment.rightHandSide().accept(*this);
-	Type const* intermediateType = type(_assignment.rightHandSide()).closestTemporaryType(
-		&type(_assignment.leftHandSide())
+
+	Token assignmentOperator = _assignment.assignmentOperator();
+	Token binaryOperator =
+		assignmentOperator == Token::Assign ?
+		assignmentOperator :
+		TokenTraits::AssignmentToBinaryOp(assignmentOperator);
+	Type const* rightIntermediateType =
+		TokenTraits::isShiftOp(binaryOperator) ?
+		type(_assignment.rightHandSide()).mobileType() :
```

Comment:
Why is this the case only for `shift`?

Reply:
Shifts are not symmetric with regards to the operands, so we do not want to convert the type of the right operand to the type of the left operand.

Reply:
Are shifts the only operators with that property?

Reply:
I think exp is also special, but in a different way. Please compare with the legacy code generator when reviewing :)

Reply:
Ok!